numpy/core/src/multiarray/scalarapi.c

```diff
@@ -344,7 +344,14 @@ PyArray_FromScalar(PyObject *scalar, PyArray_Descr *outcode)
     if (PyArray_EquivTypes(outcode, typecode)) {
         if (!PyTypeNum_ISEXTENDED(typecode->type_num)
                 || (outcode->elsize == typecode->elsize)) {
-            Py_DECREF(typecode); Py_DECREF(outcode);
+            /*
+             * Since the type is equivalent, and we haven't handed the array
+             * to anyone yet, let's fix the dtype to be what was requested,
+             * even if it is equivalent to what was passed in.
+             */
+            Py_SETREF(((PyArrayObject_fields *)arr)->descr, outcode);
```

Comment:
```suggestion
            Py_SETREF(((PyArrayObject_fields *)r)->descr, outcode);
```
Not that `arr` isn't the better name :).